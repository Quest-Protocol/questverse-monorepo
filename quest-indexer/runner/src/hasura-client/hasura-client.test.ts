import type fetch from 'node-fetch';

import HasuraClient from './hasura-client';

describe('HasuraClient', () => {
  const oldEnv = process.env;

  const HASURA_ENDPOINT = 'mock-hasura-endpoint';
  const HASURA_ADMIN_SECRET = 'mock-hasura-admin-secret';
  const PGHOST = 'localhost';
  const PGPORT = '5432';

  beforeAll(() => {
    process.env = {
      ...oldEnv,
      HASURA_ENDPOINT,
      HASURA_ADMIN_SECRET,
      PGHOST,
      PGPORT,
    };
  });

  afterAll(() => {
    process.env = oldEnv;
  });

  it('creates a schema', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.createSchema('dbName', 'schemaName');

    expect(mockFetch.mock.calls).toMatchSnapshot();
  });

  it('checks if a schema exists within source', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({
          result: [['schema_name'], ['name']]
        })
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    const result = await client.doesSchemaExist('source', 'schema');

    expect(result).toBe(true);
    expect(mockFetch.mock.calls).toMatchSnapshot();
  });

  it('checks if datasource exists', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({
          metadata: {
            sources: [
              {
                name: 'name'
              }
            ]
          },
        }),
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await expect(client.doesSourceExist('name')).resolves.toBe(true);
    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();
  });

  it('runs migrations for the specified schema', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.runMigrations('dbName', 'schemaName', 'CREATE TABLE blocks (height numeric)');

    expect(mockFetch.mock.calls).toMatchSnapshot();
  });

  it('gets table names within a schema', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify([
          { name: 'table_name', schema: 'morgs_near' },
          { name: 'height', schema: 'schema' },
          { name: 'width', schema: 'schema' }
        ])
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    const names = await client.getTableNames('schema', 'source');

    expect(names).toEqual(['height', 'width']);
    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();
  });

  it('tracks the specified tables for a specified schema', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.trackTables('schema', ['height', 'width'], 'source');

    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();
  });

  it('untracks the specified tables', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.untrackTables('default', 'schema', ['height', 'width']);

    expect(mockFetch.mock.calls).toMatchSnapshot();
  });

  it('adds the specified permissions for the specified roles/table/schema', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.addPermissionsToTables('schema', 'default', ['height', 'width'], 'role', ['select', 'insert', 'update', 'delete']);

    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();
  });

  it('adds a datasource', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({})
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.addDatasource('morgs_near', 'password', 'morgs_near');

    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();
  });

  it('tracks foreign key relationships', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({
          result: [
            [
              'coalesce'
            ],
            [
              JSON.stringify([
                {
                  table_schema: 'public',
                  table_name: 'comments',
                  constraint_name: 'comments_post_id_fkey',
                  ref_table_table_schema: 'public',
                  ref_table: 'posts',
                  column_mapping: {
                    post_id: 'id'
                  },
                  on_update: 'a',
                  on_delete: 'a'
                },
                {
                  table_schema: 'public',
                  table_name: 'post_likes',
                  constraint_name: 'post_likes_post_id_fkey',
                  ref_table_table_schema: 'public',
                  ref_table: 'posts',
                  column_mapping: {
                    post_id: 'id'
                  },
                  on_update: 'a',
                  on_delete: 'c'
                }
              ])
            ]
          ]
        }),
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.trackForeignKeyRelationships('public', 'source');

    expect(mockFetch.mock.calls[0][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[0][1].body)).toMatchSnapshot();

    expect(mockFetch.mock.calls[1][1].headers['X-Hasura-Admin-Secret']).toBe(HASURA_ADMIN_SECRET);
    expect(JSON.parse(mockFetch.mock.calls[1][1].body)).toMatchSnapshot();
  });

  it('skips foreign key tracking if none exist', async () => {
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({
          result: [
            [
              'coalesce'
            ],
            [
              JSON.stringify([])
            ]
          ]
        }),
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });

    await client.trackForeignKeyRelationships('public', 'source');

    expect(mockFetch).toBeCalledTimes(1); // to fetch the foreign keys
  });

  it('returns connection parameters for valid and invalid users', async () => {
    const testUsers = {
      testA_near: 'passA',
      testB_near: 'passB',
      testC_near: 'passC'
    };
    const TEST_METADATA = generateMetadata(testUsers);
    const mockFetch = jest
      .fn()
      .mockResolvedValue({
        status: 200,
        text: () => JSON.stringify({ metadata: TEST_METADATA })
      });
    const client = new HasuraClient({ fetch: mockFetch as unknown as typeof fetch });
    const result = await client.getDbConnectionParameters('testB_near');
    expect(result).toEqual(generateConnectionParameter('testB_near', 'passB'));
    await expect(client.getDbConnectionParameters('fake_near')).rejects.toThrow('Could not find connection parameters for user fake_near on respective database.');
  });
});

function generateMetadata (testUsers: any): any {
  const sources = [];
  // Insert default source which has different format than the rest
  sources.push({
    name: 'default',
    kind: 'postgres',
    tables: [],
    configuration: {
      connection_info: {
        database_url: { from_env: 'HASURA_GRAPHQL_DATABASE_URL' },
        isolation_level: 'read-committed',
        pool_settings: {
          connection_lifetime: 600,
          idle_timeout: 180,
          max_connections: 50,
          retries: 1
        },
        use_prepared_statements: true
      }
    }
  });

  Object.keys(testUsers).forEach((user) => {
    sources.push(generateSource(user, testUsers[user]));
  });

  return {
    version: 3,
    sources
  };
}

function generateSource (user: string, password: string): any {
  return {
    name: user,
    kind: 'postgres',
    tables: [],
    configuration: {
      connection_info: {
        database_url: { connection_parameters: generateConnectionParameter(user, password) },
        isolation_level: 'read-committed',
        use_prepared_statements: false
      }
    }
  };
}

function generateConnectionParameter (user: string, password: string): any {
  return {
    database: user,
    host: 'postgres',
    password,
    port: 5432,
    username: user
  };
}
