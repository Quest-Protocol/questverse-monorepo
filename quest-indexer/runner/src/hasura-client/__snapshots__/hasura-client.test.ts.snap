// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasuraClient adds a datasource 1`] = `
{
  "args": {
    "configuration": {
      "connection_info": {
        "database_url": {
          "connection_parameters": {
            "database": "morgs_near",
            "host": "localhost",
            "password": "password",
            "port": 5432,
            "username": "morgs_near",
          },
        },
      },
    },
    "name": "morgs_near",
  },
  "type": "pg_add_source",
}
`;

exports[`HasuraClient adds the specified permissions for the specified roles/table/schema 1`] = `
{
  "args": [
    {
      "args": {
        "permission": {
          "allow_aggregations": true,
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "height",
          "schema": "schema",
        },
      },
      "type": "pg_create_select_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "height",
          "schema": "schema",
        },
      },
      "type": "pg_create_insert_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "height",
          "schema": "schema",
        },
      },
      "type": "pg_create_update_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "height",
          "schema": "schema",
        },
      },
      "type": "pg_create_delete_permission",
    },
    {
      "args": {
        "permission": {
          "allow_aggregations": true,
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "width",
          "schema": "schema",
        },
      },
      "type": "pg_create_select_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "width",
          "schema": "schema",
        },
      },
      "type": "pg_create_insert_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "width",
          "schema": "schema",
        },
      },
      "type": "pg_create_update_permission",
    },
    {
      "args": {
        "permission": {
          "check": {},
          "columns": "*",
          "computed_fields": [],
          "filter": {},
        },
        "role": "role",
        "source": "default",
        "table": {
          "name": "width",
          "schema": "schema",
        },
      },
      "type": "pg_create_delete_permission",
    },
  ],
  "type": "bulk",
}
`;

exports[`HasuraClient checks if a schema exists within source 1`] = `
[
  [
    "mock-hasura-endpoint/v2/query",
    {
      "body": "{"type":"run_sql","args":{"sql":"SELECT schema_name FROM information_schema.schemata WHERE schema_name = 'schema'","read_only":true,"source":"source"}}",
      "headers": {
        "X-Hasura-Admin-Secret": "mock-hasura-admin-secret",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`HasuraClient checks if datasource exists 1`] = `
{
  "args": {},
  "type": "export_metadata",
  "version": 2,
}
`;

exports[`HasuraClient creates a schema 1`] = `
[
  [
    "mock-hasura-endpoint/v2/query",
    {
      "body": "{"type":"run_sql","args":{"sql":"CREATE schema schemaName","read_only":false,"source":"dbName"}}",
      "headers": {
        "X-Hasura-Admin-Secret": "mock-hasura-admin-secret",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`HasuraClient gets table names within a schema 1`] = `
{
  "args": {
    "source": "source",
  },
  "type": "pg_get_source_tables",
}
`;

exports[`HasuraClient runs migrations for the specified schema 1`] = `
[
  [
    "mock-hasura-endpoint/v2/query",
    {
      "body": "{"type":"run_sql","args":{"sql":"\\n      set schema 'schemaName';\\n      CREATE TABLE blocks (height numeric)\\n      ","read_only":false,"source":"dbName"}}",
      "headers": {
        "X-Hasura-Admin-Secret": "mock-hasura-admin-secret",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`HasuraClient tracks foreign key relationships 1`] = `
{
  "args": {
    "read_only": true,
    "source": "source",
    "sql": "
      SELECT
        COALESCE(json_agg(row_to_json(info)), '[]'::JSON)
      FROM (
        SELECT
          q.table_schema::text AS table_schema,
          q.table_name::text AS table_name,
          q.constraint_name::text AS constraint_name,
          min(q.ref_table_table_schema::text) AS ref_table_table_schema,
          min(q.ref_table::text) AS ref_table,
          json_object_agg(ac.attname, afc.attname) AS column_mapping,
          min(q.confupdtype::text) AS on_update,
          min(q.confdeltype::text) AS
          on_delete
        FROM (
          SELECT
            ctn.nspname AS table_schema,
            ct.relname AS table_name,
            r.conrelid AS table_id,
            r.conname AS constraint_name,
            cftn.nspname AS ref_table_table_schema,
            cft.relname AS ref_table,
            r.confrelid AS ref_table_id,
            r.confupdtype,
            r.confdeltype,
            unnest(r.conkey) AS column_id,
            unnest(r.confkey) AS ref_column_id
          FROM
            pg_constraint r
            JOIN pg_class ct ON r.conrelid = ct.oid
            JOIN pg_namespace ctn ON ct.relnamespace = ctn.oid
            JOIN pg_class cft ON r.confrelid = cft.oid
            JOIN pg_namespace cftn ON cft.relnamespace = cftn.oid
          WHERE
            r.contype = 'f'::"char"
            AND ((ctn.nspname='public'))
            ) q
          JOIN pg_attribute ac ON q.column_id = ac.attnum
            AND q.table_id = ac.attrelid
          JOIN pg_attribute afc ON q.ref_column_id = afc.attnum
            AND q.ref_table_id = afc.attrelid
          GROUP BY
            q.table_schema,
            q.table_name,
            q.constraint_name) AS info;
      ",
  },
  "type": "run_sql",
}
`;

exports[`HasuraClient tracks foreign key relationships 2`] = `
{
  "args": [
    {
      "args": {
        "name": "comments",
        "source": "source",
        "table": {
          "name": "posts",
          "schema": "public",
        },
        "using": {
          "foreign_key_constraint_on": {
            "column": "post_id",
            "table": {
              "name": "comments",
              "schema": "public",
            },
          },
        },
      },
      "type": "pg_create_array_relationship",
    },
    {
      "args": {
        "name": "post",
        "source": "source",
        "table": {
          "name": "comments",
          "schema": "public",
        },
        "using": {
          "foreign_key_constraint_on": "post_id",
        },
      },
      "type": "pg_create_object_relationship",
    },
    {
      "args": {
        "name": "post_likes",
        "source": "source",
        "table": {
          "name": "posts",
          "schema": "public",
        },
        "using": {
          "foreign_key_constraint_on": {
            "column": "post_id",
            "table": {
              "name": "post_likes",
              "schema": "public",
            },
          },
        },
      },
      "type": "pg_create_array_relationship",
    },
    {
      "args": {
        "name": "post",
        "source": "source",
        "table": {
          "name": "post_likes",
          "schema": "public",
        },
        "using": {
          "foreign_key_constraint_on": "post_id",
        },
      },
      "type": "pg_create_object_relationship",
    },
  ],
  "type": "bulk",
}
`;

exports[`HasuraClient tracks the specified tables for a specified schema 1`] = `
{
  "args": [
    {
      "args": {
        "source": "source",
        "table": {
          "name": "height",
          "schema": "schema",
        },
      },
      "type": "pg_track_table",
    },
    {
      "args": {
        "source": "source",
        "table": {
          "name": "width",
          "schema": "schema",
        },
      },
      "type": "pg_track_table",
    },
  ],
  "type": "bulk",
}
`;

exports[`HasuraClient untracks the specified tables 1`] = `
[
  [
    "mock-hasura-endpoint/v1/metadata",
    {
      "body": "{"type":"bulk","args":[{"type":"pg_untrack_table","args":{"table":{"schema":"schema","name":"height"},"source":"default","cascade":true}},{"type":"pg_untrack_table","args":{"table":{"schema":"schema","name":"width"},"source":"default","cascade":true}}]}",
      "headers": {
        "X-Hasura-Admin-Secret": "mock-hasura-admin-secret",
      },
      "method": "POST",
    },
  ],
]
`;
